File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/OracleType.scala
package dev.typr.foundationssc

class OracleType[T](override val underlying: dev.typr.foundations.OracleType[T]) extends DbType[T](underlying):
  override def opt: OracleType[Option[T]] =
    OracleType(underlying.opt().to(Bijections.optionalToOption))

  override def to[B](bijection: dev.typr.foundations.Bijection[T, B]): OracleType[B] =
    OracleType(underlying.to(bijection))

  def transform[B](f: dev.typr.foundations.SqlFunction[T, B], g: java.util.function.Function[B, T]): OracleType[B] =
    OracleType(underlying.transform(f, g))

  def unchecked(): OracleType[T] = OracleType(underlying.unchecked())
  def nullableOk(): OracleType[T] = OracleType(underlying.nullableOk())

File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/OracleType.kt
@file:Suppress("unused")
package dev.typr.foundationskt

class OracleType<T>(override val underlying: dev.typr.foundations.OracleType<T>) : DbType<T>(underlying) {
    override fun opt(): OracleType<T?> =
        OracleType(underlying.opt().to(Bijection.optionalToNullable()))

    override fun <B> to(bijection: Bijection<T, B>): OracleType<B> =
        OracleType(underlying.to(bijection.underlying))

    fun <B> transform(f: (T) -> B, g: (B) -> T): OracleType<B> =
        OracleType(underlying.transform(dev.typr.foundations.SqlFunction { f(it) }, g))

    fun unchecked(): OracleType<T> = OracleType(underlying.unchecked())
    fun nullableOk(): OracleType<T> = OracleType(underlying.nullableOk())
}
